### DISCUSSION TITLE
STM32 ADC on Vref Channel 17 doesn't work correctly under 3.5

### DISCUSSION THREAD

**Discussion by ronsegal** [zephyrproject-rtos/zephyr#67074] [Q&A] [answered]
Could do with some assistance to resolve battery voltage monitoring ADC code that works with Zephyr 3.4 but no longer works with 3.5.

The code,shown below, which is the first part of a battery monitoring function, initially attempts to read the reference voltage from the internal reference channel 17 for calibration purposes.
No errors are thrown up during initialisation and by ADC read but the ADC value being read by getADCLevel(&sequence) is very random, clearly incorrect.
Under 3.4 exactly the same code returns a consistently correct value.
I notice that for Zephyr 3.5 the STM32L0 adc device tree property "vref-channel" has been removed (the target chip is an STM32L073, inside an AcSIP S76S SIP), so no doubt the problem is related to this change as it is the vref-channel being read.
Similar code later in the same function that reads the actual battery voltage from BAT_VOLTS_ADC_CHAN 13 is working fine, so the problem is related to reading from channel 17.

```
uint16_t getADCLevel(const struct adc_sequence *seq) {
	int ret;
	ret = adc_read(adc_dev, seq);
	if (ret < 0) {
		LOG_ERR("adc read failed: %d", ret);
		return 0;
	}
	uint8_t *buf;
	buf = (uint8_t*)seq->buffer;
	
	 return ( (uint16_t ) (buf[1] << 8 | buf[0]) );
}

  uint8_t GetBatteryVolts( void )  {

	#define SAMPLEBUFSIZE 16
	#define ADC_GAIN ADC_GAIN_1
	#define ADC_REFERENCE ADC_REF_INTERNAL
	#define ADC_ACQUISITION_TIME ADC_ACQ_TIME_DEFAULT
	#define ADC_1ST_CHANNEL_ID 13
	#define BAT_VOLTS_ADC_CHAN 13
	#define ADC_CHANNEL_VREFINTERNAL 17
	#define ADC_DIFFERENTIAL 0
	#define ADC_RESOLUTION 12

	static struct adc_channel_cfg m_1st_channel_cfg = {
	.gain = ADC_GAIN,
	.reference = ADC_REFERENCE,
	.acquisition_time = ADC_ACQUISITION_TIME,
	.channel_id = ADC_CHANNEL_VREFINTERNAL,
	.differential = ADC_DIFFERENTIAL
	};

	uint8_t  m_sample_buffer[SAMPLEBUFSIZE];

	struct adc_sequence sequence = {
	.channels = BIT(ADC_CHANNEL_VREFINTERNAL),
	.buffer = m_sample_buffer,
	.buffer_size = SAMPLEBUFSIZE,
	.resolution = ADC_RESOLUTION,
	};

	uint8_t batteryLevel = 0;
	uint16_t measured[SAMPLEBUFSIZE];
	float measuredLevel = 0;
	uint16_t upper, lower;

	int ret;


	ret = adc_channel_setup(adc_dev, &m_1st_channel_cfg);
	if (ret < 0) {
		LOG_ERR("adc channel setup failed: %d", ret);
		return 0;
	}

	ret = adc_read(adc_dev, &sequence);
	if (ret < 0) {
		LOG_ERR("adc read failed: %d", ret);
		return 0;
	}

	uint16_t *ptr =   ( uint16_t* ) ((uint32_t) 0x1FF80078U );
	uint16_t value = *ptr;
	uint32_t supplyLevelmV =  ((uint32_t) (3000 * value))/getADCLevel(&sequence);

        ... rest of function
```

dts for the board is:

```
&adc1 {
	status = "okay";
	pinctrl-0 = <&adc_in13_pc3>;
    pinctrl-1 = <&adc_in13_pc3>;
	pinctrl-names = "default", "sleep";
	st,adc-clock-source = <SYNC>;
	st,adc-prescaler = <4>;
};
```
Which in Zephyr 3.4 referenced stm32l0.dtsi with ADC as follows:

```
adc1: adc@40012400 {
			compatible = "st,stm32-adc";
			reg = <0x40012400 0x400>;
			clocks = <&rcc STM32_CLOCK_BUS_APB2 0x00000200>;
			interrupts = <12 0>;
			status = "disabled";
			#io-channel-cells = <1>;
			temp-channel = <18>;
			vref-channel = <17>;
			resolutions = <STM32_ADC_RES(12, 0x00)
				       STM32_ADC_RES(10, 0x01)
				       STM32_ADC_RES(8, 0x02)
				       STM32_ADC_RES(6, 0x03)>;
			sampling-times = <2 4 8 13 20 40 80 161>;
			num-sampling-time-common-channels = <1>;
		};
```
In 3.5 this has become:

```
		adc1: adc@40012400 {
			compatible = "st,stm32-adc";
			reg = <0x40012400 0x400>;
			clocks = <&rcc STM32_CLOCK_BUS_APB2 0x00000200>;
			interrupts = <12 0>;
			status = "disabled";
			#io-channel-cells = <1>;
			resolutions = <STM32_ADC_RES(12, 0x00)
				       STM32_ADC_RES(10, 0x01)
				       STM32_ADC_RES(8, 0x02)
				       STM32_ADC_RES(6, 0x03)>;
			sampling-times = <2 4 8 13 20 40 80 161>;
			num-sampling-time-common-channels = <1>;
			st,adc-sequencer = <NOT_FULLY_CONFIGURABLE>;
		};
```

Would appreciate a steer on how to fix the indicated problem.

PS - Comparing STM32 adc driver code between v3.4.0 and v3.5.99 it appears that for some unknown reason STM has dropped code that configures adc to read internal channels vbat, vtemp and vref (e.g. for vref setting VREFEN bit in the ADC_CCR register). No wonder there is no sensible output.  So right now it appears that a supplementary driver will need to be written.

**Comment by ronsegal** [comment, marked as answer]
Ok, so responding to my own question now it seems, which hopefully will be helpful to others.
To enable Vref output for its internal adc channel (would be the same for temp and vbat internal), there may be a better approach but in any case I did the following:
1. Prepare the application to use STM HAL by adding a Kconfig specification file to the application:

```
 mainmenu "My Application"

config MYAPP_STM32
  default y
  bool
  select USE_STM32_HAL_ADC
  select USE_STM32_HAL_ADC_EX

source "Kconfig.zephyr"
```
2. Add `#include <soc.h>` to the application main.c
3. Before adc_channel_setup in main for channel 17, internal vref add  `HAL_ADCEx_EnableVREFINT( );` then after adc_read add `HAL_ADCEx_DisableVREFINT( );`

Have no idea why STM removed the code that takes care of enabling internal ADC channels from this latest version of zephyr 3.5.99, with the above being unnecessary in zephyr 3.4.0.  Be very interested if somebody could say why  this was done.

**Reply by gautierg-st** [reply]
Hello,
The code that enabled internal ADC channels was removed due to a limitation on STM32F4 that use the same channel for temperature and Vbat. It was not possible to know which of the two we had to read with the ADC API. See #62345
The recommended solution to read the sensor channels is now to use the sensor driver stm32_temp, stm32_vbat and stm32_vref.

**Reply by ronsegal** [reply]
Thanks for the explanation.  It would be useful to have an example that shows how stm32_vref, together with factory stored VREFINT_CAL, can be used to calibrate the ADC to read external channels.

**Comment by FRASTM** [comment]
Did you run the `west build  -p always -b nucleo_l073rz samples/sensor/soc_voltage/`
with an overlay file as follows : 
```
/ {
        aliases {
                volt-sensor0 = &vref;
        };
};
&vref {
	status = "okay";
};

&adc1 {
	status = "okay";
};
```

The value should come on the console, reading channel 17:
```
*** Booting Zephyr OS build zephyr-v3.5.0-3516-gfdea8518783a ***
Sensor voltage[vref]: 3.33 V
```

**Reply by ronsegal** [reply]
Thanks.  The aim has been to return a raw ADC value from the internal Vref to calibrate the ADC against the factory stored bandgap voltage reference value (at 1FF80078U). Guess though I could compare the voltage returned by the sensor_channel_get function, as indicated in the voltage_soc example, with the bandgap reference then adjust ADC readings from other channels to compensate by the amount that's off.  Will have a play, as using the HAL as above is a bit of kludge but does work ok. Appreciate the suggestion.
